### PR TITLE
Fix foreach

### PR DESCRIPTION
**Pull request by devotox**

### CODE REVIEW

**Comment by devotox** [comment]
@toniov @robjtede @GrayedFox @N0ps32

**Review by GrayedFox — APPROVED** [review]
Seems like this brings the forEach return inline with the other methods (it returns a Promise that fulfils with void instead of returning void by itself).

Can't say I'm too familiar with type script however so would appreciate another pair of eyes :eyes:

**Comment by robjtede** [comment]
Seems reasonable except using Promise<void> as the callback return type actually has stricter requirements for the callbacks. See [this playground](https://www.typescriptlang.org/play?ts=4.0.2#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVJxgFEowALAeQmAB4AVAPgAoAoed+WGKATwC54dANoBdADRsOYKBAgAjUgGsBTMMhhxUGAGozkIAXTHwsqUAA8BqZAFs5IGMa69DogJTwAvA3gA3HFjAEhzwGGRYAM4AgjAA5gD8AlCoPCxuAgAKMDg2kSA0-oEMANwsLKCQsAgo6Nh4BMSkZAByIADu9MyS7M78gqLBUjLySipqGiBauhD6hsamFla29o6cGi79Ih7e8Fk5eQUBwAyD7GGRMQlJKWmZ2bkR+YXHpWUA9G-wETzo8NKyCjAihYiEIJHIVGATCEAEZjAAmYwAZnE8HwO3wAGp4DC3KVQY1yK02tC4fBEfAUcZ0T4sTi8WUCeDKNRSQjkaiaWjkLIGUymsS2eSOdSvLSeRAGSCwU1IUKKVS0WL4GhQIhTCBgHyZUT2vKRUqdqqQOrUJqpSwPl8fmA-sNAYp4HAMOp8AAHe55aWEllQ2Hsymc5V7B4gAB0cAiOAgvhATDpuO1PsF-uFgdFOxDeQjICjMbjCbcUv5ENZqYVQcznseObzsfjEqL+J1LT15YNXKzNcj0fr1l5xZbcvb6cNPi74Z7+aYxtN5qTzJTZIrGfH1cnud7cdnGq1UqtUG+v3+IyB3uZw+XBsPNrHSuxuObybbV9HN9+XMLrxLvv1b6Ptqfo2T6Li+AaKu+gHKv2krfkOZavhBAF3juZpaiBApgWmSG3lyqHmqUQA) for the effect of this change.

TL;DR The current return type is fine. If anything, this PR is harmful to DX.

**Comment by salolivares** [comment]
Agree with @robjtede. However on newer versions of ESLint (w/typescript) it throws an error/warning:
```
Promise returned in function argument where a void return was expected  @typescript-eslint/no-misused-promises
```
with something like:

```typescript
import { forEach } from 'p-iteration';

await forEach(section.content, async (content) => {
    if (content.data) {
    ....
```

Wish there is a way we can type it better -- for now, I'm just disabling the ESLint rule.

**Comment by toniov** [comment]
Sorry for the late reaction, I have been out for a while.

I'm not too familiar with Typescript, but to me both ways make sense. If I have to choose I prefer to keep the flexibility we have now.

Feel free to comment here

**Comment by toniov** [comment]
Sorry for the late reaction, I have been out for a while.

I'm not too familiar with Typescript, but to me both ways make sense. If I have to choose I prefer to keep the flexibility we have now.

Feel free to comment here if you have a better idea or good arguments about this, so I can reopen the PR.

Thank you everyone!

**Comment by devotox** [comment]
The reason for the change is that you are looking for void but it is also a Promise and thus we get this issue

```
Promise returned in function argument where a void return was expected  @typescript-eslint/no-misused-promises
```

**Comment by devotox** [comment]
@toniov @salolivares @robjtede @GrayedFox what if we did this with `void | Promise<void>` ?

**Comment by GrayedFox** [comment]
Hmm, for a second I thought you meant `||` in order to short circuit the return type (haha).

I realize now you want to extend the function definition return types to include both `void` and `Promise<void>`.

The thing is by specifying `void` as the return type the code already technically allows and includes a `Promise<void>` as a possible returned type (the difference being that in the latter case, a Promise is returned that resolves to undefined as opposed to evaluating the expression and then returning undefined directly). Specifying `void` as the returned type does not preclude `Promise<void>` as once such possibility.

By it's very nature the `void` operator "[evaluates the given expression and then returns undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void)", which is in my understanding what we want here: the forEach and forEachSeries methods in the p-iteration module are designed to allow developers to use Promises (or the `await` keyword) within these prototype array methods which themselves [don't return a type](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#return_value).

Think about how we use the standard forEach method on an array i.e. open up your console and run something like this:

```
let myVar = 'suchVar';

[1, 2, 3, 4].forEach(el => console.log(el))
// logs 1, 2, 3, 4, normal usage of forEach

console.log(myVar);
// logs 'suchVar'

myVar = [1, 2, 3, 4].forEach(el => console.log(el))
// logs 1, 2, 3, 4, abnormal usage of forEach (this is a code smell)

console.log(myVar)
// logs undefined, since the returned value of forEach is always undefined
```

I could also wrap up the returned type of forEach into a Promise if I really wanted to: this still won't change the fact that forEach returns undefined and that is also what we want the p-iteration module to do. I haven't really used typescript that much but [the docs](https://www.typescriptlang.org/docs/handbook/2/functions.html) seem to encourage using void for "a function... that doesn't have a value".

Long story short: I think this is one of the few cases where the [no-misused-promises](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-misused-promises.md) rule is overreaching.

That or should we instead do the following (untested):

```
  export const forEachSeries: <T>(
    array: T[],
    callback: (currentValue: T, index: number, array: T[]) => void,
    thisArg?: any
  ) => void;
```

Does the above get rid of the lint warning and not break anything?

**Comment by GrayedFox** [comment]
This [stack overflow answer](https://stackoverflow.com/questions/59664670/typescript-async-function-return-type-void-vs-promisevoid) makes the difference between returning `void` and `Promise<void>` explicit and is pretty succinct.

File: index.d.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ declare module "p-iteration" {
    */
   export const forEachSeries: <T>(
     array: T[],
-    callback: (currentValue: T, index: number, array: T[]) => void,
+    callback: (currentValue: T, index: number, array: T[]) => Promise<void>,
     thisArg?: any
   ) => Promise<void>;
```
Comment (GrayedFox) [inline index.d.ts:29]: What about specifying void here (and line 16) as the returned type (i.e. undefined)? Does that bring everything in line and get around the eslint warning? @toniov
